File: runtime/src/buffer/tensor_view.cpp
// SPDX-License-Identifier: Apache-2.0
//
// V01-E02-F06-T01 / T02: TensorView validation, dtype/layout string
// mappings, and byte-width tables shared with the JSON Schema and the
// Rust mirror.

#include "tensorplate/buffer/tensor_view.hpp"

#include <array>
#include <cstddef>
#include <cstdint>
#include <limits>
#include <string_view>
#include <utility>

#include "tensorplate/core/error.hpp"
#include "tensorplate/core/result.hpp"

namespace tensorplate {

namespace {

constexpr std::array<std::pair<DType, std::string_view>, 9> kDtypeNames = {{
    {DType::Float32, "float32"},
    {DType::Float16, "float16"},
    {DType::BFloat16, "bfloat16"},
    {DType::Int64, "int64"},
    {DType::Int32, "int32"},
    {DType::Int16, "int16"},
    {DType::Int8, "int8"},
    {DType::UInt8, "uint8"},
    {DType::Bool, "bool"},
}};

constexpr std::array<std::pair<Layout, std::string_view>, 2> kLayoutNames = {{
    {Layout::RowMajor, "row_major"},
    {Layout::ColMajor, "col_major"},
}};

constexpr std::array<std::pair<DType, std::size_t>, 9> kDtypeByteWidths = {{
    {DType::Float32, 4},
    {DType::Float16, 2},
    {DType::BFloat16, 2},
    {DType::Int64, 8},
    {DType::Int32, 4},
    {DType::Int16, 2},
    {DType::Int8, 1},
    {DType::UInt8, 1},
    {DType::Bool, 1},
}};

}  // namespace

std::size_t dtype_byte_width(DType dtype) noexcept {
  for (const auto& [d, width] : kDtypeByteWidths) {
    if (d == dtype) {
      return width;
    }
  }
  return 0;
}

std::string_view to_string(DType dtype) noexcept {
  for (const auto& [d, name] : kDtypeNames) {
    if (d == dtype) {
      return name;
    }
  }
  return "float32";
}

std::string_view to_string(Layout layout) noexcept {
  for (const auto& [l, name] : kLayoutNames) {
    if (l == layout) {
      return name;
    }
  }
  return "row_major";
}

std::optional<DType> dtype_from_string(std::string_view name) noexcept {
  for (const auto& [d, candidate] : kDtypeNames) {
    if (candidate == name) {
      return d;
    }
  }
  return std::nullopt;
}

std::optional<Layout> layout_from_string(std::string_view name) noexcept {
  for (const auto& [l, candidate] : kLayoutNames) {
    if (candidate == name) {
      return l;
    }
  }
  return std::nullopt;
}

std::int64_t TensorView::num_elements() const noexcept {
  std::int64_t n = 1;
  for (auto d : shape_) {
    n *= d;
  }
  return n;
}

namespace {

// Computes product(shape) * dtype_byte_width(dtype) with overflow-aware
// arithmetic. Returns std::nullopt on overflow so the caller can surface a
// typed Error::Code::ShapeMismatch.
std::optional<std::size_t> compute_byte_size(const std::vector<std::int64_t>& shape,
                                             DType dtype) noexcept {
  const std::size_t element_width = dtype_byte_width(dtype);
  if (element_width == 0) {
    return std::nullopt;
  }
  std::size_t total = element_width;
  for (auto d : shape) {
    if (d <= 0) {
      return std::nullopt;
    }
    const auto ud = static_cast<std::size_t>(d);
    if (ud != 0 && total > std::numeric_limits<std::size_t>::max() / ud) {
      return std::nullopt;
    }
    total *= ud;
  }
  return total;
}

}  // namespace

Result<TensorView> TensorView::create(DType dtype, std::vector<std::int64_t> shape, Layout layout,
                                      std::size_t byte_offset, std::size_t byte_size) {
  // Enum range checks. Anything outside the declared values comes from
  // upstream JSON decoding of an unknown string; surface the typed error.
  if (dtype_byte_width(dtype) == 0) {
    return unexpected(Error::Code::ConfigInvalid, "TensorView.dtype is not a recognized value");
  }
  if (layout != Layout::RowMajor && layout != Layout::ColMajor) {
    return unexpected(Error::Code::ConfigInvalid, "TensorView.layout is not a recognized value");
  }
  if (shape.empty()) {
    return unexpected(Error::Code::ShapeMismatch, "TensorView.shape must be rank >= 1");
  }
  for (auto d : shape) {
    if (d <= 0) {
      return unexpected(Error::Code::ShapeMismatch,
                        "TensorView.shape entries must be >= 1; zero-volume tensors are not "
                        "representable in v0.1.0");
    }
  }
  auto computed = compute_byte_size(shape, dtype);
  if (!computed.has_value()) {
    return unexpected(Error::Code::ShapeMismatch,
                      "TensorView byte-size computation overflowed std::size_t");
  }
  if (byte_size == 0) {
    byte_size = *computed;
  } else if (byte_size < *computed) {
    return unexpected(Error::Code::ShapeMismatch,
                      "TensorView.byte_size is smaller than product(shape) * "
                      "dtype_byte_width(dtype)");
  }

  TensorView v;
  v.dtype_ = dtype;
  v.layout_ = layout;
  v.shape_ = std::move(shape);
  v.byte_offset_ = byte_offset;
  v.byte_size_ = byte_size;
  return v;
}

}  // namespace tensorplate

File: protocol/rust/src/tensor_view.rs
// SPDX-License-Identifier: Apache-2.0
//
// V01-E02-F06: Rust mirror of `protocol/schemas/tensor_view.json` and the
// C++ `tensorplate::TensorView` value object.

use serde::{Deserialize, Serialize};

use crate::{DecodeError, ValidatePayload, SCHEMA_VERSION};

/// Element data type. Stable wire names match the JSON Schema and the C++
/// `tensorplate::DType` enum.
#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DType {
    #[default]
    Float32,
    Float16,
    Bfloat16,
    Int64,
    Int32,
    Int16,
    Int8,
    Uint8,
    Bool,
}

impl DType {
    /// Width in bytes of one element of this dtype.
    #[must_use]
    pub fn byte_width(self) -> u64 {
        match self {
            Self::Float32 | Self::Int32 => 4,
            Self::Float16 | Self::Bfloat16 | Self::Int16 => 2,
            Self::Int64 => 8,
            Self::Int8 | Self::Uint8 | Self::Bool => 1,
        }
    }

    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Float32 => "float32",
            Self::Float16 => "float16",
            Self::Bfloat16 => "bfloat16",
            Self::Int64 => "int64",
            Self::Int32 => "int32",
            Self::Int16 => "int16",
            Self::Int8 => "int8",
            Self::Uint8 => "uint8",
            Self::Bool => "bool",
        }
    }
}

/// Memory layout. v0.1.0 supports row-major (C-contiguous) and column-major
/// (Fortran-contiguous) only; per-axis stride support is deferred.
#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Layout {
    #[default]
    RowMajor,
    ColMajor,
}

impl Layout {
    #[must_use]
    pub fn as_str(self) -> &'static str {
        match self {
            Self::RowMajor => "row_major",
            Self::ColMajor => "col_major",
        }
    }
}

/// Mirror of `tensorplate::TensorView` (C++) and
/// `protocol/schemas/tensor_view.json`.
#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct TensorView {
    pub schema_version: String,
    pub dtype: DType,
    pub shape: Vec<i64>,
    #[serde(default, skip_serializing_if = "is_default_layout")]
    pub layout: Layout,
    #[serde(default, skip_serializing_if = "is_zero")]
    pub byte_offset: u64,
    #[serde(default, skip_serializing_if = "is_zero")]
    pub byte_size: u64,
}

#[allow(clippy::trivially_copy_pass_by_ref)]
fn is_default_layout(l: &Layout) -> bool {
    *l == Layout::RowMajor
}

#[allow(clippy::trivially_copy_pass_by_ref)]
fn is_zero(v: &u64) -> bool {
    *v == 0
}

/// Errors raised by [`TensorView::new`]. Mirrors the C++
/// `TensorView::create` validation rules.
#[derive(Debug, thiserror::Error)]
pub enum TensorViewError {
    #[error("TensorView.shape must be rank >= 1")]
    EmptyShape,
    #[error("TensorView.shape entries must be >= 1; zero-volume tensors are not representable")]
    NonPositiveDim,
    #[error("TensorView.byte_size is smaller than product(shape) * dtype.byte_width()")]
    InsufficientByteSize,
    #[error("TensorView byte-size computation overflowed u64")]
    ByteSizeOverflow,
}

impl TensorView {
    /// Build and validate a TensorView at the v0.1 schema version.
    ///
    /// # Errors
    ///
    /// See [`TensorViewError`]. The same rules are applied by the C++
    /// `TensorView::create` factory.
    pub fn new(
        dtype: DType,
        shape: Vec<i64>,
        layout: Layout,
        byte_offset: u64,
        byte_size: u64,
    ) -> Result<Self, TensorViewError> {
        if shape.is_empty() {
            return Err(TensorViewError::EmptyShape);
        }
        let mut total: u64 = dtype.byte_width();
        for &d in &shape {
            if d <= 0 {
                return Err(TensorViewError::NonPositiveDim);
            }
            let ud = u64::try_from(d).map_err(|_| TensorViewError::ByteSizeOverflow)?;
            total = total
                .checked_mul(ud)
                .ok_or(TensorViewError::ByteSizeOverflow)?;
        }
        let resolved_size = if byte_size == 0 {
            total
        } else if byte_size < total {
            return Err(TensorViewError::InsufficientByteSize);
        } else {
            byte_size
        };
        Ok(Self {
            schema_version: SCHEMA_VERSION.to_string(),
            dtype,
            shape,
            layout,
            byte_offset,
            byte_size: resolved_size,
        })
    }

    /// Number of elements (product of shape).
    #[must_use]
    pub fn num_elements(&self) -> i64 {
        self.shape.iter().product()
    }
}

impl ValidatePayload for TensorView {
    fn validate_payload(self) -> Result<Self, DecodeError> {
        Self::new(
            self.dtype,
            self.shape,
            self.layout,
            self.byte_offset,
            self.byte_size,
        )
        .map_err(|err| DecodeError::InvalidPayload(err.to_string()))
    }
}

#[cfg(test)]
mod tests {
    #![allow(clippy::expect_used)]

    use super::{DType, Layout, TensorView, TensorViewError, SCHEMA_VERSION};
    use crate::decode_with_version_check;

    #[test]
    fn round_trip_preserves_fields() {
        let v = TensorView::new(DType::Float16, vec![1, 3, 224, 224], Layout::RowMajor, 0, 0)
            .expect("valid");
        let json = serde_json::to_string(&v).expect("serialize");
        let back: TensorView = serde_json::from_str(&json).expect("deserialize");
        assert_eq!(v, back);
        assert_eq!(back.schema_version, SCHEMA_VERSION);
    }

    #[test]
    fn defaults_are_omitted_for_compactness() {
        let v = TensorView::new(DType::Float32, vec![10], Layout::RowMajor, 0, 0).expect("valid");
        let json = serde_json::to_string(&v).expect("serialize");
        // Defaults compress out: layout=row_major, byte_offset=0, but byte_size
        // is computed (not zero), so it appears.
        assert!(!json.contains("\"layout\""));
        assert!(!json.contains("\"byte_offset\""));
        // byte_size = 40 (10 * 4)
        assert!(json.contains("\"byte_size\":40"));
    }

    #[test]
    fn byte_size_is_computed_when_zero() {
        let v = TensorView::new(DType::Float32, vec![2, 3], Layout::RowMajor, 0, 0).expect("valid");
        assert_eq!(v.byte_size, 24);
        assert_eq!(v.num_elements(), 6);
    }

    #[test]
    fn explicit_byte_size_must_be_at_least_computed() {
        // Padding allowed (computed=24, allocated=64).
        let v = TensorView::new(DType::Float32, vec![2, 3], Layout::RowMajor, 0, 64).expect("ok");
        assert_eq!(v.byte_size, 64);

        // Underflow rejected.
        assert!(matches!(
            TensorView::new(DType::Float32, vec![2, 3], Layout::RowMajor, 0, 16),
            Err(TensorViewError::InsufficientByteSize)
        ));
    }

    #[test]
    fn empty_shape_is_rejected() {
        assert!(matches!(
            TensorView::new(DType::Float32, vec![], Layout::RowMajor, 0, 0),
            Err(TensorViewError::EmptyShape)
        ));
    }

    #[test]
    fn non_positive_dim_is_rejected() {
        assert!(matches!(
            TensorView::new(DType::Float32, vec![1, 0, 3], Layout::RowMajor, 0, 0),
            Err(TensorViewError::NonPositiveDim)
        ));
        assert!(matches!(
            TensorView::new(DType::Float32, vec![1, -1], Layout::RowMajor, 0, 0),
            Err(TensorViewError::NonPositiveDim)
        ));
    }

    #[test]
    fn dtype_byte_width_table_is_locked() {
        assert_eq!(DType::Float32.byte_width(), 4);
        assert_eq!(DType::Float16.byte_width(), 2);
        assert_eq!(DType::Bfloat16.byte_width(), 2);
        assert_eq!(DType::Int64.byte_width(), 8);
        assert_eq!(DType::Int32.byte_width(), 4);
        assert_eq!(DType::Int16.byte_width(), 2);
        assert_eq!(DType::Int8.byte_width(), 1);
        assert_eq!(DType::Uint8.byte_width(), 1);
        assert_eq!(DType::Bool.byte_width(), 1);
    }

    #[test]
    fn version_check_decoder_accepts_current_schema() {
        let json = format!(
            r#"{{"schema_version":"{SCHEMA_VERSION}","dtype":"float16","shape":[1,3,224,224]}}"#
        );
        let v: TensorView = decode_with_version_check(&json).expect("decode");
        assert_eq!(v.dtype, DType::Float16);
        assert_eq!(v.shape, vec![1, 3, 224, 224]);
        assert_eq!(v.layout, Layout::RowMajor);
        assert_eq!(v.byte_size, 3 * 224 * 224 * 2);
    }

    #[test]
    fn version_check_decoder_rejects_old_schema() {
        let json = r#"{"schema_version":"0.0","dtype":"float32","shape":[1]}"#;
        let err = decode_with_version_check::<TensorView>(json).expect_err("rejected");
        assert!(matches!(
            err,
            crate::DecodeError::UnsupportedSchemaVersion { .. }
        ));
    }

    #[test]
    fn version_check_decoder_rejects_current_schema_invalid_shape() {
        let json =
            format!(r#"{{"schema_version":"{SCHEMA_VERSION}","dtype":"float32","shape":[1,0]}}"#);
        let err = decode_with_version_check::<TensorView>(&json).expect_err("rejected");
        assert!(matches!(err, crate::DecodeError::InvalidPayload(_)));
    }
}
